ext/drwrap/drwrap.h

```diff
@@ -88,34 +88,41 @@ drwrap_exit(void);
 DR_EXPORT
 /**
  * When #drwrap_global_flags_t #DRWRAP_INVERT_CONTROL is set, the user
- * must call this function from a drmgr analysis event handler
+ * must call this function from a drmgr insertion event handler
  * (typically registered with
  * drmgr_register_bb_instrumentation_event() or if using drbbdup with
- * the analyze_case_ex field in #drbbdup_options_t).  It is up to the user
- * to control the ordering, since the priority #DRMGR_PRIORITY_INSERT_DRWRAP
- * will not apply.
+ * the instrument_instr_ex field in #drbbdup_options_t).  This
+ * function will insert instrumentation for function wrapping pre and
+ * post callbacks.  It is up to the user to control the ordering,
+ * since the priority #DRMGR_PRIORITY_INSERT_DRWRAP will not apply.
+ * The separate "where" handles cases such as with drbbdup's final app
+ * instruction (which cannot be duplicated into each case) or with
+ * emulation where the instruction "inst" to monitor is distinct from
+ * the location "where" to insert instrumentation.
  */
 dr_emit_flags_t
-drwrap_invoke_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
-                       bool translating, OUT void **user_data);
+drwrap_invoke_insert(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst,
+                     instr_t *where, bool for_trace, bool translating, void *user_data);
 
 DR_EXPORT
 /**
- * When #drwrap_global_flags_t #DRWRAP_INVERT_CONTROL is set, the user
- * must call this function from a drmgr insertion event handler
- * (typically registered with
- * drmgr_register_bb_instrumentation_event() or if using drbbdup with
- * the instrument_instr_ex field in #drbbdup_options_t).  It is up to the user
- * to control the ordering, since the priority #DRMGR_PRIORITY_INSERT_DRWRAP
- * will not apply.
+ * When #drwrap_global_flags_t #DRWRAP_INVERT_CONTROL is set and the
+ * user is using multiple types of instrumentation such as with
+ * drbbdup, the user must call this function for each instrumentation
+ * case where function wrapping should not be enabled.  This will
+ * insert required cleanup for instrumentation cases changing in the
+ * middle of a wrapped function.  It will not invoke any wrapped
```

Comment:
Maybe add that there's an opportunity for drbbdup to switch instrumentation  modes because the post-call callback is in a different fragment.

Reply:
I don't follow?  Do you mean a different drbbdup case, rather than a different fragment?

Reply:
Just describing the problem in some more detail:

The drbbdup instrumentation case changes happen at the fragment boundary. As cleanup is part of a different fragment, the instrumentation case may change before we can complete cleanup.

Reply:
Feel free to ignore if being more verbose doesn't add value here.

Reply:
If the case changes to don't-wrap and then changes back to wrap all inside a wrapped callee: the full post-wrap callback plus cleanup would then be called.  I think that is what we want.

Reply:
Adding a sentence about that.